marconi-sidechain-experimental/src/Marconi/Sidechain/Experimental/Api/Types.hs

```diff
@@ -0,0 +1,50 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module Marconi.Sidechain.Experimental.Api.Types (
+  Types.configPort,
+  Types.configTrace,
+  module Marconi.Sidechain.Experimental.Api.Types,
+) where
+
+import Cardano.Api qualified as C
+import Control.Lens (makeLenses, (^.))
+import Control.Monad.Except (ExceptT, mapExceptT)
+import Control.Monad.Reader (ReaderT, withReaderT)
+import Data.List.NonEmpty (NonEmpty)
+import Marconi.ChainIndex.Api.Types qualified as Types
+
+{- SERVER CONFIG -}
+
+{- | 'SidechainHttpServerConfig' augments @Types.'HttpServerConfig'@ with
+ - 'SidechainExtraHttpServerConfig', containing anything needed for sidechain
+ - queries not included in HttpServerConfig.
+-}
+data SidechainHttpServerConfig = SidechainHttpServerConfig
+  { _chainIndexHttpServerConfig :: !Types.HttpServerConfig
+  , _sidechainExtraHttpServerConfig :: !SidechainExtraHttpServerConfig
+  }
+
+{- SIDECHAIN EXTRA CONFIG -}
+
+type TargetAssets = NonEmpty (C.PolicyId, Maybe C.AssetName)
+
+newtype SidechainExtraHttpServerConfig = SidechainExtraHttpServerConfig
+  { _sidechainTargetAssets :: Maybe TargetAssets
+  }
+
+{- LENSES -}
+makeLenses ''SidechainExtraHttpServerConfig
+makeLenses ''SidechainHttpServerConfig
+
+{- UTILITIES -}
+
+-- | Specialization of 'withReaderT'.
+withChainIndexHttpServerConfig
+  :: ReaderT Types.HttpServerConfig m a -> ReaderT SidechainHttpServerConfig m a
+withChainIndexHttpServerConfig = withReaderT (^. chainIndexHttpServerConfig)
+
+-- | Specialization of 'mapError' and 'withReaderT' in the case used here.
+mapChainIndexExceptT
+  :: ExceptT e (ReaderT Types.HttpServerConfig m) a
+  -> ExceptT e (ReaderT SidechainHttpServerConfig m) a
+mapChainIndexExceptT = mapExceptT withChainIndexHttpServerConfig
```

Comment:
I'm not sure this is the most helpful name for the function. (Naming is hard!) It suggests it's doing something with `ExceptT` whereas really it's doing something with the reader environment.

How about `useChainIndexHandler` or `withChainIndexHandler`? Given how it's used, I think that would express its purpose rather than its implementation.

I also think `withChainIndexHttpServerConfig` might be renamed `withChainIndexReaderT`. However, even better might be to remove it and put its definition inline in the (renamed) `mapChainIndexExceptT`.

I wonder too if using the `ReaderHandler` type alias would be more consistent rather than writing out the types in full. Having the full types helps the reader see why `mapExceptT` and `withReaderT` are being used, but I think it's fairly obvious once you see that they are being used.

I also noticed that `ReaderHandler` and `ReaderServer` are defined in two places, `marconi-core-json-rpc/src/Marconi/Core/JsonRpc.hs` and `marconi-sidechain/src/Marconi/Sidechain/Api/HttpServer.hs`. I think the latter should get them from the former. Also, that makes me think the main part of `mapChainIndexExceptT` should perhaps be moved to `Marconi.Core.JsonRpc` and generalized, something like this:

```haskell
withReaderHandler :: (r' -> r) -> ReaderHandler r a -> ReaderHandler r' a
withReaderHandler = mapExceptT . withReaderT
```

Then there could be a convenience function here that uses it:

```haskell
withChainIndexHandler :: ReaderHandler HttpServerConfig a -> ReaderHandler SidechainHttpServerConfig a
withChainIndexHandler = withReaderHandler (^. chainIndexHttpServerConfig)
```

Reply:
I think adding a class for `HasHttpServerConfig` might be over-engineering it, at least until we see what the future use cases might be like. I feel like adding a typeclass is a lot more noise and cognitive overhead than adding a per-environment helper function, partly because the class needs to be defined globally. I'm also not convinced that `HttpServerConfig` is a general enough type to warrant having its own typeclass.

Down the road, it might be appropriate to consider adding multiple `Has` classes for the items in the config that handlers would need access to, so that we have more fine-grained descriptions of what the handlers are actually using. It's then simple to add the instances for the derived environment using lens composition, eg

```
instance HasTrackedAddresses env where
  trackedAddressesL = baseEnvL . trackedAddressesL
```

For now, however, I think it's best just to keep using `withChainIndexHandler` but I think refactoring it to use a `withReaderHandler` in `marconi-core-json-rpc` is such a small-but-helpful change that it's appropriate to include it in this PR.

Reply:
Thanks for the comments. I've added `withReaderHandler` to `marconi-core-json-rpc` and updated `withChainIndexHandler` to use it.